servicex/query_cache.py

```diff
@@ -69,25 +69,75 @@ def transformed_results(self, transform: TransformRequest,
         )
 
     def cache_transform(self, record: TransformedResults):
+        transforms = Query()
         with self.lock:
-            if not self.contains_hash(record.hash):
-                self.db.insert(json.loads(record.model_dump_json()))
+            self.db.upsert(json.loads(record.model_dump_json()), transforms.hash == record.hash)
 
     def update_record(self, record: TransformedResults):
         transforms = Query()
         with self.lock:
             self.db.update(json.loads(record.model_dump_json()), transforms.hash == record.hash)
 
     def contains_hash(self, hash: str) -> bool:
+        """
+        Check if the cache has completed records for a hash
+        """
         transforms = Query()
         with self.lock:
-            records = self.db.search(transforms.hash == hash)
+            records = self.db.search((transforms.hash == hash) & (transforms.files.exists()))
```

Comment:
I *strongly* prefer this test is on `transforms.status != 'SUBMITTED'`. As it is, we are relying on the fact that certain information is not in a record to infer the type of the record, when we actually do have the relevant field!

Reply:
I guess maybe it's `~(transforms.status == 'SUBMITTED')` ...